### PR TITLE
[jit] improve error message on recursive class defs

### PR DESCRIPTION
**Pull request by suo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#21842 [jit] improve error message on recursive class defs**
* #21841 [jit] improve recursive scripting error message

We don't support this yet, so it should error clearly.

Differential Revision: [D15857568](https://our.internmc.facebook.com/intern/diff/D15857568)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -16268,15 +16268,30 @@ def copy(self):
...
-                # type: (Tree) -> None
-                child.parent = torch.jit.annotate(Optional[Tree], self)
+                # type: (Leaf) -> None
+                self.child = child
+
+    def test_recursive_class(self):
+        """
+        Recursive class types not yet supported. We should give a good error message.
+        """
+        with self.assertRaisesRegex(RuntimeError, "Recursive class attributes not yet supported"):
```
Comment (zdevito) [inline test/test_jit.py]: "Recursive class" is a bit vague as a concept can it be made more clear what is being objected to here?

**Review by zdevito — APPROVED** [review]
Yeah, good idea.

File: torch/csrc/jit/script/sugared_value.cpp [hunk 2]
```diff
@@ -148,6 +148,21 @@ std::vector<std::shared_ptr<SugaredValue>> SimpleValue::asTuple(
...
 
+static bool isRecursive(const TypePtr& classType, const TypePtr& attrType) {
+  if (attrType->isSubtypeOf(classType)) {
+    return true;
+  }
+
+  // Recursively check contained types. We need to do this because a user may do
+  // A -> B -> A.
+  for (const auto& type : attrType->containedTypes()) {
+    if (isRecursive(classType, type)) {
```
Comment (eellison) [inline torch/csrc/jit/script/sugared_value.cpp]: The naive search here seems like it could get really inefficient with any complicated hierarchies

Comment (suo) [reply torch/csrc/jit/script/sugared_value.cpp]: BFS would be faster given how types are structured. But like, how complicated are we talking lol—this search is linear in the size of the type DAG, so unless someone is making 100000-nested compound types I think we're ok.

Comment (suo) [reply torch/csrc/jit/script/sugared_value.cpp]: I guess we're doing a second traversal if it's a tuple type, so if we have deep nested tuples we may need to add manually unrolling.

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/script/sugared_value.cpp (hunk 2 above)
Comment (eellison) [reply torch/csrc/jit/script/sugared_value.cpp]: I think this is exponential in the worst case...

Comment (eellison) [reply torch/csrc/jit/script/sugared_value.cpp]: If you have `N` numbered nodes where each node is connected to all nodes with a lower number

**Pull request closed without merging**